sys/include/net/netdev2.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2015 Ell-i open source co-operative
+ *               2015 Freie Universität Berlin
+ *               2014 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @defgroup    net_netdev2 Low-Level Network Driver Interface
+ * @ingroup     net_netdev2
+ * @brief       Low-level network driver interface
+ * @{
+ *
+ * This interface is supposed to be a low-level interface for networkdrivers.
+ * @file
+ * @brief       Definitions low-level network driver interface
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Martine Lenders <mlenders@inf.fu-berlin.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef NETDEV2_H
+#define NETDEV2_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "ng_netconf.h"
+#include "netdev_events.h"
+
+/**
+ * @brief   Event callback for signaling event to upper layers
+ *
+ * @param[in] type          type of the event
+ * @param[in] arg           event argument
+ */
+typedef int (*netdev2_event_cb_t)(netdev2_t *dev, netdev2_event_t event, void* arg);
+
+/**
+ * @brief Structure to hold driver state
+ *
+ * Supposed to be extended by driver implementations.
+ * The extended structure should contain all variable driver state.
+ */
+typedef struct netdev2 {
+    const struct netdev2_driver *driver;    /**< ptr to that driver's interface. */
+    netdev2_event_cb_t event_callback;      /**< callback for device events */
+} netdev2_t;
+
+/**
+ * @brief Structure to hold driver interface -> function mapping
+ *
+ * The send/receive functions expect/return a full ethernet
+ * frame (dst mac, src mac, ethertype, payload, no checksum).
+ */
+typedef struct netdev2_driver {
+    /**
+     * @brief Send frame
+     *
+     * @param[in] dev       network device descriptor
+     * @param[in] vector    io vector array to send
+     * @param[in] count     nr of entries in vector
+     *
+     * @return nr of bytes sent, or <=0 on error
+     */
+    int (*send)(netdev2_t *dev, const struct iovec *vector, int count);
```

Comment:
> Wouldn't it make sense to streamline the return value for error cases to always use negative numbers?

Probably, but recv is supposed to be called whenever a packet is received, and if it then returns "0", something is wrong. So mentioning here that 0 is also an error might save implementors the pain of accidently expecting a raw packet with size==0.